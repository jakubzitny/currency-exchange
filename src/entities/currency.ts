// export type Currency = 'USD' | 'GBP' | 'EUR'

export enum Currency {
  USD = 'USD',
  GBP = 'GBP',
  EUR = 'EUR',
}
