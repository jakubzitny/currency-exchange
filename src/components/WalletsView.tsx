import { List } from 'immutable'
import React, { useEffect, useState } from 'react'
import useSWR from 'swr'

import Arrow from './Arrow'
import WalletView from './WalletView'
import { Currency } from '../entities/currency'
import Wallet from '../entities/wallet'
import { roundAmount, validateAmount } from '../utils/validation-utils'

import styles from '../styles/WalletsView.module.css'

type Props = {
  rate: number
  wallets: List<Wallet>
  sourceWalletCurrency: Currency
  targetWalletCurrency: Currency
  onSourceWalletChange: (currency: Currency) => void
  onTargetWalletChange: (currency: Currency) => void
}

type State = {
  sourceWallet: Wallet
  targetWallet: Wallet
}

const WalletsView: React.FC<Props> = (props: Props) => {
  const { sourceWalletCurrency, targetWalletCurrency, wallets } = props
  const sourceWallet = wallets.find(
    (wallet) => wallet.get('currency') === sourceWalletCurrency
  )
  const targetWallet = wallets.find(
    (wallet) => wallet.get('currency') === targetWalletCurrency
  )
  if (!sourceWallet || !targetWallet) {
    throw new Error('Wallet misconfiguration!')
  }

  const [amountState, setAmountState] = useState({
    exchangeAmount: 50,
    exchangedAmount: props.rate ? roundAmount(50 * props.rate) : 67.22,
  })

  useEffect(() => {
    setAmountState((prevState) => ({
      ...prevState,
      exchangedAmount: roundAmount(props.rate * prevState.exchangeAmount),
    }))
  }, [props.rate])

  const handleSourceAmountChange = (amount: number) => {
    const exchangedAmount = roundAmount(amount * props.rate)
    setAmountState({ exchangeAmount: amount, exchangedAmount })
  }

  const handleTargetAmountChange = (amount: number) => {
    const exchangeAmount = roundAmount(amount / props.rate)
    setAmountState({ exchangeAmount, exchangedAmount: amount })
  }

  return (
    <div className={styles.container}>
      <WalletView
        exchangeAmount={amountState.exchangeAmount}
        exchangeCurrency={targetWallet.currency}
        exchangeRate={props.rate}
        wallet={sourceWallet}
        wallets={wallets}
        type="source"
        onAmountChange={handleSourceAmountChange}
        onWalletChange={props.onSourceWalletChange}
      />
      <Arrow />
      <WalletView
        exchangeAmount={amountState.exchangedAmount}
        exchangeCurrency={sourceWallet.currency}
        exchangeRate={props.rate}
        wallet={targetWallet}
        wallets={wallets}
        type="target"
        onAmountChange={handleTargetAmountChange}
        onWalletChange={props.onTargetWalletChange}
      />
    </div>
  )
}

export default WalletsView
