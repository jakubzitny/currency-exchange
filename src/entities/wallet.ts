import { Record } from 'immutable'
import { Currency } from './currency'

interface Defaults {
  currency: Currency
  amount: number
}

const defaults: Defaults = {
  currency: Currency.EUR,
  amount: 0,
}

export default class Wallet extends Record<Defaults>(defaults) {}
