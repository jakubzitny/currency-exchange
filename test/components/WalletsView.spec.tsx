import chai, { expect } from 'chai'
import chaiSinon from 'sinon-chai'
import chaiJestSnapshot from 'chai-jest-snapshot'
import { shallow } from 'enzyme'
import React from 'react'
import { spy } from 'sinon'
import { List } from 'immutable'

import { Currency } from '../../src/entities/currency'
import Wallet from '../../src/entities/wallet'
import WalletsView from '../../src/components/WalletsView'

chai.use(chaiJestSnapshot)
chai.use(chaiSinon)

describe('WalletsView', () => {
  const createWalletsView = (props: Object = {}) => {
    const wallet = new Wallet({ currency: Currency.EUR, amount: 0 })
    const wallet2 = new Wallet({ currency: Currency.USD, amount: 0 })
    const defaultProps = {
      rate: 1,
      wallets: List.of(wallet, wallet2),
      sourceWalletCurrency: Currency.EUR,
      targetWalletCurrency: Currency.USD,
      onSourceWalletChange: () => {},
      onTargetWalletChange: () => {},
    }
    const switcher = <WalletsView {...defaultProps} {...props} />
    return shallow(switcher)
  }

  it('should render WalletsView', () => {
    const wrapper = createWalletsView()
    expect(wrapper).to.matchSnapshot()
  })

  it('should handle source wallet change', () => {
    const onSourceWalletChange = spy()
    const wrapper = createWalletsView({ onSourceWalletChange })

    const props: any = wrapper.find('WalletView').at(0).props()
    props.onWalletChange(Currency.USD)

    expect(onSourceWalletChange).to.have.been.calledOnceWith(Currency.USD)
  })

  it('should handle target wallet change', () => {
    const onTargetWalletChange = spy()
    const wrapper = createWalletsView({ onTargetWalletChange })

    const props: any = wrapper.find('WalletView').at(1).props()
    props.onWalletChange(Currency.USD)

    expect(onTargetWalletChange).to.have.been.calledOnceWith(Currency.USD)
  })

  it('should handle source amount change and adapt exchange amounts accordingly', () => {
    const rate = 1.5
    const wrapper = createWalletsView({ rate })

    const props: any = wrapper.find('WalletView').at(0).props()
    props.onAmountChange(50)

    const updatedSourceWalletViewProps: any = wrapper
      .find('WalletView')
      .at(0)
      .props()
    const updatedTargetWalletViewProps: any = wrapper
      .find('WalletView')
      .at(1)
      .props()

    expect(updatedSourceWalletViewProps.exchangeAmount).to.equal(50)
    expect(updatedTargetWalletViewProps.exchangeAmount).to.equal(75)
  })

  it('should handle source amount change and adapt exchange amounts accordingly', () => {
    const rate = 0.8
    const wrapper = createWalletsView({ rate })

    const props: any = wrapper.find('WalletView').at(1).props()
    props.onAmountChange(50)

    const updatedSourceWalletViewProps: any = wrapper
      .find('WalletView')
      .at(0)
      .props()
    const updatedTargetWalletViewProps: any = wrapper
      .find('WalletView')
      .at(1)
      .props()

    expect(updatedSourceWalletViewProps.exchangeAmount).to.equal(62.5)
    expect(updatedTargetWalletViewProps.exchangeAmount).to.equal(50)
  })
})
