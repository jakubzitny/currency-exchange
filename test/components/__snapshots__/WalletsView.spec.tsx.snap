// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WalletsView should render WalletsView 1`] = `ShallowWrapper {}`;
